### PR TITLE
Add params to machinepool and support instance configuration upgrade

### PR DESCRIPTION
**Pull request by shyamradhakrishnan**
**What this PR does / why we need it**:

Add params to machinepool and support instance configuration upgrade, and add supported parameters.

The vendor folder was removed via commit https://github.com/oracle/cluster-api-provider-oci/commit/bf3d86c5afb6bd2097a9b9e42336a73eab8ad998 so that the vendor Custom resources are not published via crddev

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

### CODE REVIEW

File: cloud/scope/machine_pool.go [hunk 1]
```diff
@@ -536,27 +596,24 @@ func (m *MachinePoolScope) CreateInstancePool(ctx context.Context) (*core.Instan
...
 	if instancePoolNeedsUpdates(m, instancePool) {
-		m.Info("updating instance pool")
-
-		replicas := int(1)
-		if m.MachinePool.Spec.Replicas != nil {
-			replicas = int(*m.MachinePool.Spec.Replicas)
-		}
-
+		m.Info("Updating instance pool")
+		replicas := int(*m.MachinePool.Spec.Replicas)
```
Comment (joekr) [inline cloud/scope/machine_pool.go:602]: Is it possible for `Replicas` at this point to be `nil`?

File: cloud/scope/machine_pool.go [hunk 2]
```diff
@@ -536,27 +596,24 @@ func (m *MachinePoolScope) CreateInstancePool(ctx context.Context) (*core.Instan
...
-
-		if _, err := m.ComputeManagementClient.UpdateInstancePool(ctx, req); err != nil {
-			return errors.Wrap(err, "unable to update instance pool")
+				Size:                    common.Int(replicas),
+				InstanceConfigurationId: m.OCIMachinePool.Spec.InstanceConfiguration.InstanceConfigurationId,
+			},
 		}
+		resp, err := m.ComputeManagementClient.UpdateInstancePool(ctx, req)
+		if err != nil {
+			return nil, err
```
Comment (joekr) [inline cloud/scope/machine_pool.go:611]: If it fails to update, I don't recall, is the error message clear enough that update failed? I think that is why the extra `Wrap` was used. 🤷

File: cloud/scope/machine_pool.go [hunk 3]
```diff
@@ -571,17 +628,313 @@ func (m *MachinePoolScope) TerminateInstancePool(ctx context.Context, instancePo
...
 		}
-	} else if instancePool.Size != nil {
-		return true
+		if len(agentConfigSpec.PluginsConfig) > 0 {
+			pluginConfigList := make([]core.InstanceAgentPluginConfigDetails, len(agentConfigSpec.PluginsConfig))
+			for i, pluginConfigSpec := range agentConfigSpec.PluginsConfig {
+				pluginConfigRequest := core.InstanceAgentPluginConfigDetails{
+					Name: pluginConfigSpec.Name,
+				}
+				desiredState, _ := core.GetMappingInstanceAgentPluginConfigDetailsDesiredStateEnum(string(pluginConfigSpec.DesiredState))
```
Comment (joekr) [inline cloud/scope/machine_pool.go:657]: If the value isn't found should the `bool` be checked error? or set a default or something?

File: cloud/scope/machine_pool.go [hunk 4]
```diff
@@ -571,17 +628,313 @@ func (m *MachinePoolScope) TerminateInstancePool(ctx context.Context, instancePo
...
+		for _, nsg := range m.OCICluster.Spec.NetworkSpec.Vcn.NetworkSecurityGroups {
+			if nsg.Role == infrastructurev1beta1.WorkerRole {
+				nsgs = append(nsgs, *nsg.ID)
+			}
+		}
+		return nsgs
+	}
+}
+
+func (m *MachinePoolScope) GetInstanceConfiguration(ctx context.Context) (*core.InstanceConfiguration, error) {
```
Comment (joekr) [inline cloud/scope/machine_pool.go:842]: Please add the go doc comments for exported functions.

**Review by joekr — CHANGES_REQUESTED** [review]

File: cloud/scope/machine.go [hunk 5]
```diff
@@ -537,7 +537,7 @@ func (m *MachineScope) ReconcileCreateInstanceOnLB(ctx context.Context) error {
 // ReconcileDeleteInstanceOnLB checks to make sure the instance is part of a backend set then deletes the backend
 // on the NetworkLoadBalancer
 //
-// It will await the Work Request completion before returning
+// # It will await the Work Request completion before returning
```
[the remark this answers is not in the document]
Comment (shyamradhakrishnan) [reply cloud/scope/machine.go]: removed

File: cloud/scope/machine_pool.go [hunk 6]
```diff
@@ -342,89 +339,143 @@ func (m *MachinePoolScope) GetFreeFormTags() map[string]string {
...
-			m.Info("failed to create instance configuration due to shape config")
 			return err
 		}
-		if (shapeConfig != core.InstanceConfigurationLaunchInstanceShapeConfigDetails{}) {
-			launchInstanceDetails.LaunchDetails.ShapeConfig = &shapeConfig
-		}
+		return m.PatchObject(ctx)
+	} else {
+		computeDetails, ok := instanceConfiguration.InstanceDetails.(core.ComputeInstanceDetails)
+		launchDetailsActual := computeDetails.LaunchDetails
```
[the remark this answers is not in the document]
Comment (shyamradhakrishnan) [reply cloud/scope/machine_pool.go]: fixed

File: cloud/scope/machine_pool.go (hunk 1 above)
Comment (shyamradhakrishnan) [reply cloud/scope/machine_pool.go]: fixed

File: cloud/scope/machine_pool.go (hunk 2 above)
Comment (shyamradhakrishnan) [reply cloud/scope/machine_pool.go]: fixed

File: cloud/scope/machine_pool.go (hunk 3 above)
Comment (shyamradhakrishnan) [reply cloud/scope/machine_pool.go]: fixed

File: cloud/scope/machine_pool.go (hunk 4 above)
Comment (shyamradhakrishnan) [reply cloud/scope/machine_pool.go:856]: fixed

**Comment by shyamradhakrishnan** [comment]
unit tests
```
Using cached envtest tools from /Users/shyaradh/oke/src/github.com/oracle/cluster-api-provider-oci/testbin
setting up env vars
?   	github.com/oracle/cluster-api-provider-oci	[no test files]
ok  	github.com/oracle/cluster-api-provider-oci/api/v1beta1	1.070s	coverage: 17.9% of statements
?   	github.com/oracle/cluster-api-provider-oci/cloud/config	[no test files]
ok  	github.com/oracle/cluster-api-provider-oci/cloud/ociutil	0.730s	coverage: 25.0% of statements
ok  	github.com/oracle/cluster-api-provider-oci/cloud/scope	168.397s	coverage: 76.7% of statements
?   	github.com/oracle/cluster-api-provider-oci/cloud/scope/mocks	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/base	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/base/mock_base	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/compute	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/compute/mock_compute	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/computemanagement	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/computemanagement/mock_computemanagement	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/containerengine	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/containerengine/mock_containerengine	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/identity	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/identity/mock_identity	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/networkloadbalancer	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/networkloadbalancer/mock_nlb	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/vcn	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/vcn/mock_vcn	[no test files]
ok  	github.com/oracle/cluster-api-provider-oci/controllers	28.468s	coverage: 69.0% of statements
ok  	github.com/oracle/cluster-api-provider-oci/exp/api/v1beta1	3.403s	coverage: 8.8% of statements
ok  	github.com/oracle/cluster-api-provider-oci/exp/controllers	1.783s	coverage: 55.3% of statements
?   	github.com/oracle/cluster-api-provider-oci/feature	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/version	[no test files]
```

**Comment by shyamradhakrishnan** [comment]
machinepool e2e test run
```
  Timeline >>
  STEP: Creating namespace "capoci-e2e-8zf93w" for hosting the cluster @ 12/06/22 10:33:58.641
  Dec  6 10:33:58.641: INFO: starting to create namespace for hosting the "capoci-e2e-8zf93w" test spec
  INFO: Creating namespace capoci-e2e-8zf93w
  INFO: Creating event watcher for namespace "capoci-e2e-8zf93w"
  INFO: Cluster name is capoci-e2e-8zf93w-machine-pool
  INFO: Creating the workload cluster with name "capoci-e2e-8zf93w-machine-pool" using the "machine-pool" template (Kubernetes v1.24.4, 1 control-plane machines, 1 worker machines)
  INFO: Getting the cluster template yaml
  INFO: clusterctl config cluster capoci-e2e-8zf93w-machine-pool --infrastructure (default) --kubernetes-version v1.24.4 --control-plane-machine-count 1 --worker-machine-count 1 --flavor machine-pool
  INFO: Applying the cluster template yaml to the cluster
  INFO: Waiting for the cluster infrastructure to be provisioned
  STEP: Waiting for cluster to enter the provisioned phase @ 12/06/22 10:34:00.996
  INFO: Waiting for control plane to be initialized
  INFO: Waiting for the first control plane machine managed by capoci-e2e-8zf93w/capoci-e2e-8zf93w-machine-pool-control-plane to be provisioned
  STEP: Waiting for one control plane node to exist @ 12/06/22 10:35:31.056
  INFO: Installing a CNI plugin to the workload cluster
  INFO: Waiting for control plane to be ready
  INFO: Waiting for control plane capoci-e2e-8zf93w/capoci-e2e-8zf93w-machine-pool-control-plane to be ready (implies underlying nodes to be ready as well)
  STEP: Waiting for the control plane to be ready @ 12/06/22 10:48:23.739
  STEP: Checking all the control plane machines are in the expected failure domains @ 12/06/22 10:49:23.77
  INFO: Waiting for the machine deployments to be provisioned
  INFO: Waiting for the machine pools to be provisioned
  STEP: Waiting for the machine pool workload nodes @ 12/06/22 10:49:23.829
  STEP: Scaling the machine pool up @ 12/06/22 10:52:53.918
  INFO: Patching the replica count in Machine Pool capoci-e2e-8zf93w/capoci-e2e-8zf93w-machine-pool-mp-0
  STEP: Waiting for the machine pool workload nodes @ 12/06/22 10:52:53.963
  STEP: Scaling the machine pool down @ 12/06/22 10:57:54.083
  INFO: Patching the replica count in Machine Pool capoci-e2e-8zf93w/capoci-e2e-8zf93w-machine-pool-mp-0
  STEP: Waiting for the machine pool workload nodes @ 12/06/22 10:57:54.147
  STEP: Dumping logs from the "capoci-e2e-8zf93w-machine-pool" workload cluster @ 12/06/22 10:59:44.194
  STEP: Dumping all the Cluster API resources in the "capoci-e2e-8zf93w" namespace @ 12/06/22 10:59:44.194
  STEP: Deleting all clusters in the capoci-e2e-8zf93w namespace @ 12/06/22 10:59:44.453
  STEP: Deleting cluster capoci-e2e-8zf93w-machine-pool @ 12/06/22 10:59:44.484
  INFO: Waiting for the Cluster capoci-e2e-8zf93w/capoci-e2e-8zf93w-machine-pool to be deleted
  STEP: Waiting for cluster capoci-e2e-8zf93w-machine-pool to be deleted @ 12/06/22 10:59:44.496
  STEP: Deleting namespace used for hosting the "create-workload-cluster" test spec @ 12/06/22 11:04:34.607
  INFO: Deleting namespace capoci-e2e-8zf93w
  << Timeline
------------------------------
[SynchronizedAfterSuite] PASSED [0.838 seconds]
[SynchronizedAfterSuite]
```

File: cloud/scope/machine_pool.go [hunk 7]
```diff
@@ -342,89 +339,146 @@ func (m *MachinePoolScope) GetFreeFormTags() map[string]string {
...
-		} else {
-			return errors.Wrap(err, fmt.Sprintf("error getting instance configuration by id %s", instanceConfigurationId))
+	instanceConfiguration, err := m.GetInstanceConfiguration(ctx)
+	if err != nil {
+		return err
+	}
+	freeFormTags := m.GetFreeFormTags()
+	definedTags := make(map[string]map[string]interface{})
+	if m.OCICluster.Spec.DefinedTags != nil {
+		for ns, mapNs := range m.OCICluster.Spec.DefinedTags {
```
Comment (joekr) [inline cloud/scope/machine_pool.go:349]: I know it is a short lived scope so no need to change but what does the N in `mapNs` stand for?

**Review by joekr — APPROVED** [review]
lgtm

**Pull request merged**